api/tls_context.proto

```diff
@@ -39,6 +41,9 @@ message TlsCertificate {
   DataSource password = 3;
   DataSource ocsp_staple = 4;
   repeated DataSource signed_certificate_timestamp = 5;
+
+  // SDS configuration which allows to fetch and/or reload this TLS certificates.
+  SdsSecretConfig sds_secret_config = 6;
```

Comment:
I'm wondering if we can just take existing sites where we have `TlsCertificate` and make them a `oneof` with `TlsCertificate` as exists today and `SdsSecretConfig`? Not sure if the rules for wire compatibility allow this if we preserve the existing field numbering. Can you look into this and see if it's possible? If so, we could avoid this weirdness (or make use of `oneof` internally within `TlsCertificate` to at least make clear these are mutually exclusive).

Reply:
Unfortunately, we can't, since there is no such thing as `repeated oneof` and existing code uses `repeated TlsCertificate`.

If we want to add `oneof` in order to enforce mutually exclusive behavior, then we would need to add something like this:

    message TlsCertificateSource {
      oneof source {
        TlsCertificate tls_certificate = 1;
        SdsSecretConfig sds_secret_config = 2;
      }
    }
    repeated TlsCertificateSource tls_certificate_sources = 6;

and deprecate existing `repeated TlsCertificate tls_certificates` (or use it as a fallback).

Alternative and backward-compatible way of adding SDS would be to add `repeated SdsSecretConfig sds_secret_config` along existing `repeated TlsCertificate tls_certificates` to the `CommonTlsContext` message, since in theory there can be a mix of static and SDS-provided certificates anyway, but that's a bit less clean API than the `oneof` approach.

Thoughts?

Reply:
I'm fine with having two `repeated` fields, one for static and the other for dynamic resources. This is similar to what we do in bootstrap. I don't think anyone is relying on the `TlsCertificate` today that we know of, but there's folks who are building management servers around v2, so it might be bad form to break them given these APIs are nominally frozen, so the `TlsCertificateSource` option probably isn't optimal form that perspective.